### ISSUE TITLE
awx 1.0.6: "NameError: name 'Queue' is not defined"

### ISSUE THREAD

**Issue by tjyang** [MrMEEE/awx-build#17]
Hi @MrMEEE 

The dev repo is now hosting  1.0.6.0-1 version of AWX rpm.
Do you know how to resolve issue like "NameError: name 'Queue' is not defined" ?

```
[root@ansible04 ~]# sudo -u awx /opt/awx/bin/awx-manage makemigrations
Traceback (most recent call last):
  File "/opt/awx/bin/awx-manage", line 11, in <module>
    sys.exit(manage())
  File "/opt/awx/embedded/lib/python2.7/site-packages/awx/__init__.py", line 93, in manage
    prepare_env()
  File "/opt/awx/embedded/lib/python2.7/site-packages/awx/__init__.py", line 48, in prepare_env
    if not settings.DEBUG: # pragma: no cover
  File "/opt/awx/embedded/lib64/python2.7/site-packages/django/conf/__init__.py", line 56, in __getattr__
    self._setup(name)
  File "/opt/awx/embedded/lib64/python2.7/site-packages/django/conf/__init__.py", line 41, in _setup
    self._wrapped = Settings(settings_module)
  File "/opt/awx/embedded/lib64/python2.7/site-packages/django/conf/__init__.py", line 110, in __init__
    mod = importlib.import_module(self.SETTINGS_MODULE)
  File "/usr/lib64/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/opt/awx/embedded/lib/python2.7/site-packages/awx/settings/production.py", line 95, in <module>
    include(settings_file, optional(settings_files), scope=locals())
  File "/opt/awx/embedded/lib/python2.7/site-packages/split_settings/tools.py", line 102, in include
    exec(compile(to_compile.read(), included_file, 'exec'), scope)
  File "/etc/awx/settings.py", line 35, in <module>
    CELERY_QUEUES += (Queue(CLUSTER_HOST_ID, Exchange(CLUSTER_HOST_ID), routing_key=CLUSTER_HOST_ID),)
NameError: name 'Queue' is not defined
[root@ansible04 ~]#

```

**Comment by MrMEEE** [comment]
Try to comment out this line in /etc/awx/settings.py:

CELERY_QUEUES += (Queue(CLUSTER_HOST_ID, Exchange(CLUSTER_HOST_ID), routing_key=CLUSTER_HOST_ID),)

Now I get a migration error:

Traceback (most recent call last):
  File "/opt/awx/bin/awx-manage", line 11, in <module>
    sys.exit(manage())
  File "/opt/awx/embedded/lib/python2.7/site-packages/awx/__init__.py", line 109, in manage
    execute_from_command_line(sys.argv)
  File "/opt/awx/embedded/lib64/python2.7/site-packages/django/core/management/__init__.py", line 364, in execute_from_command_line
    utility.execute()
  File "/opt/awx/embedded/lib64/python2.7/site-packages/django/core/management/__init__.py", line 356, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/opt/awx/embedded/lib64/python2.7/site-packages/django/core/management/base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/opt/awx/embedded/lib64/python2.7/site-packages/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/opt/awx/embedded/lib64/python2.7/site-packages/django/core/management/commands/makemigrations.py", line 96, in handle
    loader = MigrationLoader(None, ignore_no_migrations=True)
  File "/opt/awx/embedded/lib64/python2.7/site-packages/django/db/migrations/loader.py", line 52, in __init__
    self.build_graph()
  File "/opt/awx/embedded/lib64/python2.7/site-packages/django/db/migrations/loader.py", line 274, in build_graph
    raise exc
django.db.migrations.exceptions.NodeNotFoundError: Migration main.0026_auto_20180305_2318 dependencies reference nonexistent parent node (u'main', u'0025_v330_delete_authtoken')

I'm not really strong in the migration stuff.. so hope I can get some help from the awx devs...

**Comment by MrMEEE** [comment]
Just got an answer from the AWX Team:

"Upgrades between AWX versions are not expected to work.  However, we have recently added an import/export capability to tower-cli/awx-cli, which allows you to export your job templates and other objects (not including credential secrets) to a JSON file, which you can then re-import to a freshly installed 1.0.6."

They are referring to the awx-cli tool from their separate repo.. also, the awx-manage tool have a dumpdata/loaddata tool...

I'm going to see if I can do a workaround for upgrades.. 

Right now.. DON'T Upgrade..

**Comment by MrMEEE** [comment]
I'm working on something like this (if you want to try, not testet proberly):

awx-cli receive --credential all > credentials
awx-cli receive --all > data
awx-cli receive --project all > project

systemctl stop awx-celery-worker awx-cbreceiver awx-celery-beat awx-channels-worker awx-daphne awx-web
su - postgres -c "dropdb awx"
su - postgres -c "createdb -O awx awx"

sudo -u awx /opt/awx/bin/awx-manage migrate

echo "from django.contrib.auth.models import User; User.objects.create_superuser('admin', 'root@localhost', 'test')" | sudo -u awx /opt/awx/bin/awx-manage shell
sudo -u awx /opt/awx/bin/awx-manage provision_instance --hostname=$(hostname)
sudo -u awx /opt/awx/bin/awx-manage register_queue --queuename=tower --hostnames=$(hostname)
awx-cli send credentials
awx-cli send project
awx-cli send data

**Comment by palachu** [comment]
Hi @MrMEEE 

I'm having the same problem both from fresh installation or upgrading to anything higher than 1.0.5.32

**Comment by MrMEEE** [comment]
@palachu: have you tried commenting out the Line:  
CELERY_QUEUES += (Queue(CLUSTER_HOST_ID, Exchange(CLUSTER_HOST_ID), routing_key=CLUSTER_HOST_ID),)

In settings.py??

**Comment by unixfox** [comment]
A fresh install of awx 1.0.6 doesn't work.
I tried to comment the line @MrMEEE but that gives me another error:
```
Traceback (most recent call last):
  File "/opt/awx/bin/awx-manage", line 11, in <module>
    sys.exit(manage())
  File "/opt/awx/embedded/lib/python2.7/site-packages/awx/__init__.py", line 109, in manage
    execute_from_command_line(sys.argv)
  File "/opt/awx/embedded/lib64/python2.7/site-packages/django/core/management/__init__.py", line 364, in execute_from_command_line
    utility.execute()
  File "/opt/awx/embedded/lib64/python2.7/site-packages/django/core/management/__init__.py", line 338, in execute
    django.setup()
  File "/opt/awx/embedded/lib64/python2.7/site-packages/django/__init__.py", line 27, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/opt/awx/embedded/lib64/python2.7/site-packages/django/apps/registry.py", line 116, in populate
    app_config.ready()
  File "/opt/awx/embedded/lib/python2.7/site-packages/awx/conf/apps.py", line 25, in ready
    configure_external_logger(settings)
  File "/opt/awx/embedded/lib/python2.7/site-packages/awx/main/utils/handlers.py", line 357, in configure_external_logger
    is_enabled = settings_module.LOG_AGGREGATOR_ENABLED
  File "/opt/awx/embedded/lib/python2.7/site-packages/awx/conf/settings.py", line 479, in __getattr_without_cache__
    return getattr(self._wrapped, name)
  File "/opt/awx/embedded/lib/python2.7/site-packages/awx/conf/settings.py", line 392, in __getattr__
    value = self._get_local(name)
  File "/usr/lib64/python2.7/contextlib.py", line 35, in __exit__
    self.gen.throw(type, value, traceback)
  File "/opt/awx/embedded/lib/python2.7/site-packages/awx/conf/settings.py", line 66, in _log_database_error
    if get_tower_migration_version() < '310':
  File "/opt/awx/embedded/lib/python2.7/site-packages/awx/main/utils/db.py", line 14, in get_tower_migration_version
    loader = MigrationLoader(connection, ignore_no_migrations=True)
  File "/opt/awx/embedded/lib64/python2.7/site-packages/django/db/migrations/loader.py", line 52, in __init__
    self.build_graph()
  File "/opt/awx/embedded/lib64/python2.7/site-packages/django/db/migrations/loader.py", line 274, in build_graph
    raise exc
django.db.migrations.exceptions.NodeNotFoundError: Migration main.0028_auto_20180404_1518 dependencies reference nonexistent parent node (u'main', u'0027_v330_add_tower_verify')
```

**Comment by MrMEEE** [comment]
I think that you might have left some _auto_ files in the migration folder...???

Or did you do a complete reinstall of the server??

Try removing all _auto_ and _merge_ files in the migration folder

**Comment by unixfox** [comment]
I did as usual:
```
sudo -u awx /opt/awx/bin/awx-manage makemigrations
sudo -u awx /opt/awx/bin/awx-manage migrate
```
After upgrading to 1.0.6 (I didn't know that the upgrade was broken).

Now I purged & reinstalled awx because I didn't have anything on it and I reinstalled an older version of awx (1.0.5). Do you want me to upgrade a second time to 1.0.6 for testing the upgrade?

**Comment by MrMEEE** [comment]
It would be great if you could try the updated upgrade guide, to verify that it works... Maybe add some entries in awx first...

**Comment by unixfox** [comment]
How do I get `awx-cli`? From here: https://github.com/NarrativeTeam/awx-cli?

**Comment by MrMEEE** [comment]
I used this:

https://github.com/ansible/tower-cli

I will push a rpm soon, probably later today...

**Comment by MrMEEE** [comment]
I have just pushed a SRPM to the dev repo.. it's currently building and should be available soon...

yum install ansible-tower-cli

**Comment by MrMEEE** [comment]
ansible-tower-cli should be available now :)

**Comment by unixfox** [comment]
Thank you @MrMEEE, I installed it manually but I appreciate your help :).

I got a working AWX with the version 1.0.6 by following your guide and your advice https://github.com/MrMEEE/awx-build/issues/17#issuecomment-385076675:

```
awx-cli receive -u user -p password --credential all > credentials
awx-cli receive -u user -p password --all > data
awx-cli receive -u user -p password --project all > project
systemctl stop awx-celery-worker awx-cbreceiver awx-celery-beat awx-channels-worker awx-daphne awx-web
su - postgres -c "dropdb awx"
su - postgres -c "createdb -O awx awx"
yum update -y
sudo -u awx /opt/awx/bin/awx-manage migrate
echo "from django.contrib.auth.models import User; User.objects.create_superuser('user', 'root@localhost', 'password')" | sudo -u awx /opt/awx/bin/awx-manage shell
sudo -u awx /opt/awx/bin/awx-manage provision_instance --hostname=$(hostname)
sudo -u awx /opt/awx/bin/awx-manage register_queue --queuename=tower --hostnames=$(hostname)
systemctl start awx-celery-worker awx-cbreceiver awx-celery-beat awx-channels-worker awx-daphne awx-web
```

For importing the previously exported data I got an error on the command `awx-cli send -u user -p password project`:
```
Failed to resolve organization Default for project Demo Project
```

**Comment by MrMEEE** [comment]
In which order did you import???

**Comment by unixfox** [comment]
```
awx-cli send credentials
```
then 
```
awx-cli send project
```

**Comment by MrMEEE** [comment]
The guide will probably need to change the export to:

awx-cli receive --organization all --team all --credential_type all  --credential all  --notification_template all --user all --inventory_script all  --inventory all  --project all  --job_template all  --workflow all > alldata

and

import to:

awx-cli send alldata

**Comment by palachu** [comment]
Thanks guys I got it working by commenting out that line and cleaning old migrations.

**Comment by MrMEEE** [comment]
This is definitely not the best upgrade path... as far as I can see i'm loosing:

Users (export is blank), and therefore user permissions
Log/History
Inventory Groups
Credential passwords
LDAP/Auth config

Would be easier if the migration stuff would work...

**Comment by wdhowe** [comment]
Just wanted to add my confirmation here.

Fresh install of the latest awx rpm (1.0.6.1-1) on a new CentOS 7.4 system worked correctly AFTER commenting out the CELERY_QUEUES line.

/etc/awx/settings.py

#CELERY_QUEUES += (Queue(CLUSTER_HOST_ID, Exchange(CLUSTER_HOST_ID), routing_key=CLUSTER_HOST_ID),)

Thank you!

**Comment by MrMEEE** [comment]
Original issue is fixed (workaround, and fixed in 1.0.6.3).. the broken upgrades/migrations has been moved to #18 ...
